tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java

```diff
@@ -0,0 +1,1513 @@
+        });
+        Map<TopicPartition, KafkaFuture<Void>> results = adminClient.alterPartitionReassignments(args).values();
+        Map<TopicPartition, Throwable> errors = new HashMap<>();
+        for (Map.Entry<TopicPartition, KafkaFuture<Void>> e :  results.entrySet()) {
+            try {
+                e.getValue().get();
+            } catch (ExecutionException t) {
+                errors.put(e.getKey(), t.getCause());
+            }
+        }
+        return errors;
+    }
+
+    /**
+     * Cancel the given partition reassignments.
+     *
+     * @param adminClient       The admin client object to use.
+     * @param reassignments     The partition reassignments to cancel.
+     * @return                  A map from partition objects to error strings.
+     */
+    static Map<TopicPartition, Throwable> cancelPartitionReassignments(Admin adminClient,
+                                                                       Set<TopicPartition> reassignments) throws InterruptedException {
+        Map<TopicPartition, Optional<NewPartitionReassignment>> args = new HashMap<>();
+        reassignments.forEach(part -> args.put(part, Optional.empty()));
+
+        Map<TopicPartition, KafkaFuture<Void>> results = adminClient.alterPartitionReassignments(args).values();
+        Map<TopicPartition, Throwable> errors = new HashMap<>();
+        for (Map.Entry<TopicPartition, KafkaFuture<Void>> e :  results.entrySet()) {
+            try {
+                e.getValue().get();
+            } catch (ExecutionException t) {
+                errors.put(e.getKey(), t.getCause());
+            }
+        }
+        return errors;
+    }
+
+    /**
+     * Compute the in progress partition move from the current reassignments.
+     * @param currentReassignments All replicas, adding replicas and removing replicas of target partitions
+     */
+    private static Map<String, Map<Integer, PartitionMove>> calculateCurrentMoveMap(Map<TopicPartition, PartitionReassignment> currentReassignments) {
+        Map<String, Map<Integer, PartitionMove>> moveMap = new HashMap<>();
+        // Add the current reassignments to the move map.
+        currentReassignments.entrySet().forEach(e -> {
+            TopicPartition part = e.getKey();
+            PartitionReassignment reassignment = e.getValue();
+
+            List<Integer> allReplicas = reassignment.replicas();
+            List<Integer> addingReplicas = reassignment.addingReplicas();
+
+            // The addingReplicas is included in the replicas during reassignment
+            Set<Integer> sources = new HashSet<>(allReplicas);
+            sources.removeAll(addingReplicas);
+
+            Set<Integer> destinations = new HashSet<>(addingReplicas);
+
+            Map<Integer, PartitionMove> partMoves = moveMap.computeIfAbsent(part.topic(), k -> new HashMap<>());
+            partMoves.put(part.partition(), new PartitionMove(sources, destinations));
+        });
```

Comment:
```suggestion
 currentReassignments.forEach((part, reassignment) -> {
            List<Integer> allReplicas = reassignment.replicas();
            List<Integer> addingReplicas = reassignment.addingReplicas();

            // The addingReplicas is included in the replicas during reassignment
            Set<Integer> sources = new HashSet<>(allReplicas);
            addingReplicas.forEach(sources::remove);

            Set<Integer> destinations = new HashSet<>(addingReplicas);

            Map<Integer, PartitionMove> partMoves = moveMap.computeIfAbsent(part.topic(), k -> new HashMap<>());
            partMoves.put(part.partition(), new PartitionMove(sources, destinations));
        });
```
Also there is recommendation to avoid `mySet.removeAll(myList);`. Fixed at the suggestion by  `myList.forEach(mySet::remove);`

Reply:
Fixed